File: benchmarks/fixtures/sources.py
"""Seeded per-language source generators for the analyzer benchmarks.

Committed generators, not committed blobs: the micro suite materializes these into a
scratch directory at run time, so the corpus scales (20/200/2000 lines) without a
megabyte of checked-in filler. Deterministic: same (language, lines, seed) → same
bytes. The exact line count is honored — analyzer cost curves are plotted against it.
"""

from __future__ import annotations

import random

LANGS = ("python", "shell", "js", "ts")
EXTENSIONS = {"python": "py", "shell": "sh", "js": "js", "ts": "ts"}

_WORDS = ("alpha", "bravo", "delta", "gamma", "kilo", "lima", "omega", "sigma")


def generate(lang: str, lines: int, seed: int = 20260720) -> str:
    """A syntactically valid source of exactly `lines` lines (plus final newline),
    carrying the constructs each language's analyzer actually looks at — parameters,
    env-defaults, argument parsing — so warm-parse timings exercise real paths."""
    if lang not in LANGS:
        raise ValueError(f"unknown language {lang!r} (expected one of {LANGS})")
    if lines < 8:
        raise ValueError("need at least 8 lines for the fixed scaffold")
    rng = random.Random(f"{seed}:{lang}:{lines}")  # noqa: S311 — deterministic fixtures, not crypto
    body = {"python": _python, "shell": _shell, "js": _js, "ts": _ts}[lang](lines, rng)
    if len(body) != lines:
        raise AssertionError(f"generator produced {len(body)} lines, wanted {lines}")
    return "\n".join(body) + "\n"


def _pad(body: list[str], lines: int, comment: str) -> list[str]:
    while len(body) < lines:
        body.append(f"{comment} filler line {len(body) + 1}")
    return body


def _python(lines: int, rng: random.Random) -> list[str]:
    body = [
        "import argparse",
        "",
        "parser = argparse.ArgumentParser(description='generated bench source')",
        f"parser.add_argument('--{rng.choice(_WORDS)}', type=int, default={rng.randrange(9)})",
        "parser.add_argument('--verbose', action='store_true')",
        "args = parser.parse_args()",
    ]
    while len(body) < lines - 2:
        word = rng.choice(_WORDS)
        body += [f"def fn_{len(body)}(x: int) -> int:", f"    return x + {len(word)}"]
    return _pad(body, lines, "#")


def _shell(lines: int, rng: random.Random) -> list[str]:
    body = ["#!/usr/bin/env bash", "set -euo pipefail"]
    body.extend(
        f'{word.upper()}="${{{word.upper()}:-{rng.randrange(99)}}}"'
        for word in _WORDS[: rng.randrange(3, 6)]
    )
    while len(body) < lines - 1:
        body.append(f'echo "step {len(body)}: ${_WORDS[len(body) % len(_WORDS)].upper()}"')
    return _pad(body, lines, "#")


def _js(lines: int, rng: random.Random) -> list[str]:
    body = [
        "const args = process.argv.slice(2);",
        f"let {rng.choice(_WORDS)} = {rng.randrange(9)};",
    ]
    # Emit only WHOLE function chunks, then pad — truncating a chunk would cut its
    # closing brace and hand the analyzer an error-recovery parse tree, silently
    # under-reporting analyze times ~5x (this actually happened; the validity test
    # in tests/test_benchmarks_tooling.py now pins it).
    while len(body) + 3 <= lines:
        body += [f"function fn{len(body)}(x) {{", f"  return x + {len(body)};", "}"]
    return _pad(body, lines, "//")


def _ts(lines: int, rng: random.Random) -> list[str]:
    body = [
        "const args: string[] = process.argv.slice(2);",
        f"let {rng.choice(_WORDS)}: number = {rng.randrange(9)};",
    ]
    # Whole chunks only — see _js.
    while len(body) + 3 <= lines:
        body += [
            f"function fn{len(body)}(x: number): number {{",
            f"  return x + {len(body)};",
            "}",
        ]
    return _pad(body, lines, "//")

File: crates/skit-benchmarks/src/sources.rs
//! Deterministic source workloads for parser benchmarks.

use thiserror::Error;

use crate::python_random::PythonRandom;

/// Languages in the stable analyzer grid.
pub const LANGUAGES: &[&str] = &["python", "shell", "js", "ts"];
const WORDS: &[&str] = &[
    "alpha", "bravo", "delta", "gamma", "kilo", "lima", "omega", "sigma",
];
const SEED: u64 = 20_260_720;

/// Source generation failed.
#[derive(Clone, Debug, Eq, Error, PartialEq)]
pub enum SourceError {
    /// The language is outside the benchmark grid.
    #[error("unknown language {0:?} (expected python, shell, js, or ts)")]
    UnknownLanguage(String),
    /// The requested source cannot hold the fixed scaffold.
    #[error("source workloads need at least 8 lines")]
    TooShort,
}

/// Return the fixture extension for one benchmark language.
#[must_use]
pub fn extension(language: &str) -> &'static str {
    match language {
        "python" => "py",
        "shell" => "sh",
        "js" => "js",
        "ts" => "ts",
        _ => "",
    }
}

/// Generate one valid parser workload with an exact line count and final newline.
pub fn generate(language: &str, lines: usize) -> Result<String, SourceError> {
    if lines < 8 {
        return Err(SourceError::TooShort);
    }
    let mut rng = PythonRandom::seeded(&format!("{SEED}:{language}:{lines}"));
    let mut body = match language {
        "python" => python(lines, &mut rng),
        "shell" => shell(lines, &mut rng),
        "js" => javascript(lines, &mut rng, false),
        "ts" => javascript(lines, &mut rng, true),
        _ => return Err(SourceError::UnknownLanguage(language.to_owned())),
    };
    let comment = if matches!(language, "js" | "ts") {
        "//"
    } else {
        "#"
    };
    while body.len() < lines {
        body.push(format!("{comment} filler line {}", body.len() + 1));
    }
    debug_assert_eq!(body.len(), lines);
    Ok(format!("{}\n", body.join("\n")))
}

/// Generate the same workload with a half-written final construct.
pub fn generate_broken(language: &str, lines: usize) -> Result<String, SourceError> {
    let truncation = match language {
        "python" => "def fn_half_written(x: int,",
        "shell" => "if [ \"${ALPHA}\" = ",
        "js" => "function fnHalfWritten(x) {",
        "ts" => "function fnHalfWritten(x: number): number {",
        _ => return Err(SourceError::UnknownLanguage(language.to_owned())),
    };
    let source = generate(language, lines)?;
    let mut body = source.lines().map(str::to_owned).collect::<Vec<_>>();
    *body.last_mut().expect("minimum line count") = truncation.to_owned();
    Ok(format!("{}\n", body.join("\n")))
}

fn python(lines: usize, rng: &mut PythonRandom) -> Vec<String> {
    let word = random_word(rng);
    let default = rng.range(0, 9);
    let mut body = vec![
        "import argparse".to_owned(),
        String::new(),
        "parser = argparse.ArgumentParser(description='generated bench source')".to_owned(),
        format!("parser.add_argument('--{word}', type=int, default={default})"),
        "parser.add_argument('--verbose', action='store_true')".to_owned(),
        "args = parser.parse_args()".to_owned(),
    ];
    while body.len() < lines - 2 {
        let length = random_word(rng).len();
        body.extend([
            format!("def fn_{}(x: int) -> int:", body.len()),
            format!("    return x + {length}"),
        ]);
    }
    body
}

fn shell(lines: usize, rng: &mut PythonRandom) -> Vec<String> {
    let mut body = vec![
        "#!/usr/bin/env bash".to_owned(),
        "set -euo pipefail".to_owned(),
    ];
    let variables = rng.range(3, 6);
    for word in &WORDS[..variables] {
        body.push(format!(
            "{}=\"${{{}:-{}}}\"",
            word.to_ascii_uppercase(),
            word.to_ascii_uppercase(),
            rng.range(0, 99)
        ));
    }
    while body.len() < lines - 1 {
        let variable = WORDS[body.len() % WORDS.len()].to_ascii_uppercase();
        body.push(format!("echo \"step {}: ${variable}\"", body.len()));
    }
    body
}

fn javascript(lines: usize, rng: &mut PythonRandom, typed: bool) -> Vec<String> {
    let word = random_word(rng);
    let default = rng.range(0, 9);
    let mut body = if typed {
        vec![
            "const args: string[] = process.argv.slice(2);".to_owned(),
            format!("let {word}: number = {default};"),
        ]
    } else {
        vec![
            "const args = process.argv.slice(2);".to_owned(),
            format!("let {word} = {default};"),
        ]
    };
    while body.len() + 3 <= lines {
        let index = body.len();
        if typed {
            body.extend([
                format!("function fn{index}(x: number): number {{"),
                format!("  return x + {index};"),
                "}".to_owned(),
            ]);
        } else {
            body.extend([
                format!("function fn{index}(x) {{"),
                format!("  return x + {index};"),
                "}".to_owned(),
            ]);
        }
    }
    body
}

fn random_word(rng: &mut PythonRandom) -> &'static str {
    rng.choice(WORDS)
}
